airbrus.mt

```diff
@@ -68,6 +68,55 @@ def UTF8JSON :DeepFrozen := composeCodec(UTF8, JSON)
 def makeLineTube() as DeepFrozen:
     return makePumpTube(makeSplitPump(b`$\n`))
 
+def make_giftExchange() :Any as DeepFrozen:
+    def givers  := [].asMap().diverge()
+    def wanters := [].asMap().diverge()
+        
+    def give(giver :Str, acceptor :Str, giftName :Str, gift :Any) :Void :
+        def giftBox := gift  # Allways wrap your presents! (For when gifts are promises)
+        if (wanters.contains(acceptor):
+            def wants := wanters[acceptor]
+            if (wants.contains(giftName)):
+              wants.fetch(giftName)[1].resolve(&giftBox)
+              return
+        if (!givers.contains(giver)):
+            givers[giver] := [].asMap().diverge()
+        def gifts := givers[giver]
+        gifts[giftName] := &giftBox
+        return
+  
+    def accept(acceptor :Str, giver :Str, giftName :Str) :Any :
+        def promres := Ref.make_promise()
+        var [promise, resolver] := promres
+        if (givers.contains(giver)):
+            def gifts := givers[giver]
+            if (gifts.contains(giftName)):
+                resolver.resolve(gifts.fetch(giftName)
+                return promise
+        if (wanters.contains(acceptor)):
+            def wants := wanters[acceptor]
+            if (wants.contains(giftName)):
+                [promise, resolver] = wants.fetch(giftName)
+                return promise
+        if (!wanters.contains(acceptor)):
+            wanters.set(acceptor, [].asMap().diverge())
+        def wants := wanters.fetch(acceptor)
+        wants.set(giftName, promres)
+        return promise
```

Comment:
Since all branches of control flow return `promise`, consider refactoring to have fewer returns.